Deere 2.1 part 2

Be-ing opened the pull request:
Continuing work on the Deere skin. Now that I have been using it for a while, I am focusing on simplifying it in this PR.

I will be refactoring it to remove the bottom splitter that has been confusing users. Having two splitters in the same direction makes resizing them wonky at times. The only purpose of that was to make it possible to hide the EQ knob labels and have the mixer take less vertical space. Instead of making that optional, I'll just remove the EQ knob labels. LateNight and Tango already do not have the EQ knobs labeled. This may be a little bit confusing for new users completely unfamiliar with DJ equipment, but so are many other parts of the GUI and tooltips are good enough for those. To anyone who has ever used any DJ gear, it shouldn't be an issue at all.

TODO:
  * [x] Remove bottom splitter
  * [x] Simplify samplers
  * [x] Add booth gain knob
  * [x] Show recording & broadcasting icons when not active
  * [ ] Polish mixer and deck sizes in all configurations
  * [x] Left click to unfocus effects
  * [x] Find a better place for vinyl control buttons?
  * [x] Find a better place for beatgrid editing buttons?

Be-ing committed:
Deere: hide headphone controls when no headphones configured

Be-ing committed:
Deere: add Booth gain knob

Be-ing committed:
Deere: cleanup skin.xml a little

daschuer commented:
Good Idea!

daschuer commented on res/skins/Deere/tool_bar.xml:
```diff
@@ -103,6 +107,28 @@
+              </Template>
+
+              <Template src="skin:knob_toolbar.xml">
+                <SetVariable name="TooltipId">master_gain</SetVariable>
```
Wrong tooltip

daschuer commented on res/skins/Deere/tool_bar.xml:
```diff
@@ -103,6 +107,28 @@
+                <SetVariable name="group">[Master]</SetVariable>
+                <SetVariable name="control">booth_gain</SetVariable>
+                <SetVariable name="color">red</SetVariable>
+                <SetVariable name="label">Booth</SetVariable>
```
Now we have Gain / Head Gain / Booth .. three naming schemas for similar features.

JosepMaJAZ replied:
Mmm... haven't checked it at all, but here is my opinion:
"Gain", without any other label, should be used where the context limits it to a specific control. I.e. track gain, sampler gain,  Master gain should tell that it is the Master.
"Head gain": if not using a hardware control (present in many controllers) the UI specific control for headphones should be labeled this way.
"Booth gain": AFAIK, the booth is the speakers/monitors that the DJ can have inside the cabin to listen to the audio. They are of importance when the cabin's position can't give the DJ enough detail. It can serve two purposes: to better equalize, or to better mix (getting a cleaner source of what's being played while also listening to the headphones). Some other times it might simply be for DJ amusement :P

Anyway, booth gain does not have much sense on Mixxx since we don't have an explicit Booth output and it should not be confused with headphones output. Booth is generally connected to the secondary output of a Mixer table, containing also the master output.

Be-ing replied:
I agree that "GAIN" without any context is not clear. I have changed it to "MASTER".

The booth output was introduced in #1279.

Be-ing committed:
Deere: change "GAIN" label to "MASTER" for master gain knob

and add tooltip for booth gain

Be-ing committed:
Deere: change "HEAD GAIN" label to just "HEAD"

Be-ing committed:
Deere: remove bottom splitter and EQ knob labels

Users were getting confused by the bottom splitter and having two
spltiters in the same direction made resizing them wonky. The only
purpose of the bottom splitter was to hide the EQ knob labels, so just
remove the EQ knob labels always.

Be-ing committed:
Deere: mixer fixes

There's some weird asymmetry going on in the mixer. Not sure why...

daschuer commented:
some time ago, we decided for the word "gain" to state clear that this is a digital pre-amplifier gain with a bitcrusher nature, useful to level the output for maximum signal noise ratio. 
It is not a volume knob controlling a digital hardware pot, which is the expectation of most users. 
As a general advise we can say "Don't touch"

If we drop the word gain, the issue is back. Can you add a Group label "Gain" for them?  

Maybe it is time to work on this bug, to finally fix this annoying and long standing issue: 
https://bugs.launchpad.net/mixxx/+bug/1440443
https://bugs.launchpad.net/mixxx/+bug/1440675
https://bugs.launchpad.net/mixxx/+bug/1458677

Be-ing commented:
> If we drop the word gain, the issue is back.

Whether it's labeled "MASTER" or "GAIN" in the skin is irrelevant to that issue IMO. Neither makes it clear that the knob does not affect the hardware volume control, nor do I think there is any succinct label that could make that clear. I understand your desire to try to implement proper volume control in an easy way that users wouldn't be likely to mess up, I do not think it is possible with the heterogeneity of operating systems and sound cards out there. Many sound cards do not even expose the hardware volume control to the computer. Even if they did, I think it would be futile to try to manipulate the hardware volume control reliably across each OS and all the complications of the various sound APIs on Windows. IMO the best we can do is document what the software gains do and advise not to use them in most cases.

> Maybe it is time to work on this bug, to finally fix this annoying and long standing issue

Can we please focus on finishing what we have started and getting a release out?

Be-ing commented:
The refactoring to get rid of the bottom splitter is done, but there is now some strange asymmetry between the left and right sides of the mixer I am trying to resolve. It is most apparent by right clicking the crossfader to center it and seeing that it does not align with the master level meter.

daschuer commented:
We had a condense about the GAIN label.  An it is an issue for me that the controls are looking like volume in Tango.

daschuer commented:
> Can we please focus on finishing what we have started and getting a release out?

We can IMHO release at any time. If we wait for all started things  to be finished (like we did now for more than year) we will get never a release. The today master is IMHO stable enough to be a beta. 
My vote was to release the last years September version now we have September again.

Be-ing committed:
Deere: fix mixer asymmetry

Be-ing committed:
Deere: adjust some borders

Be-ing commented:
I am thinking about removing the "Minimal Controls" mode. I implemented that as a quick hack around the clutter of the old looping UI. Now that all controls fit into a reasonable area, I think it can be removed. I have found myself frustrated and confused a number of times when I had the Minimal Controls mode active and wanted access to the hidden controls. Thoughts?

Be-ing committed:
Deere: simplify some file names

Be-ing committed:
Deere: simplify sampler templates

Be-ing committed:
Deere: simplify sampler expanded view

This reduces shutdown time by about 1 second for me.

daschuer commented:
> I am thinking about removing the "Minimal Controls" mode.

That would be Ok for me. We can advertise Deere as a full feature skin. Tango and Schade can be focused on other use cases. An additional Broadcasting skin would be real great

Be-ing committed:
Deere: remove crossfader orientation button from mics

Microphones can't be assigned to the crossfader in the engine.

Be-ing committed:
Deere: remove minimal deck controls mode

This was a quick hack around the clutter of the old looping UI. Now that
all controls fit into a reasonable amount of space and the bottom
splitter has been removed, get rid of this option. On many occasions I
found myself confused why I couldn't access a control in minimal
controls mode then annoyed that I had to open the skin settings menu to
get to what I wanted.

Be-ing committed:
Deere: remove QuickEffect text labels

TODO: Make a new knob widget to show the proper tooltip

Be-ing committed:
Deere: make QuickEffect kill switches match EQ kill switches

Be-ing committed:
Deere: add option to hide EQ kill switches

Be-ing committed:
Deere: let mixer shrink vertically

Be-ing commented:
I added an option to hide the EQ kill switches. Few controllers have buttons for those so they waste space for most users. However, they're useful for mouse & keyboard users.

ronso0 commented:
> We had a condense about the GAIN label. An it is an issue for me that the controls are looking like volume in Tango.

I wasn't aware of that. Guess you refer to controls in mixer bar? I have absolutely no problem changing them to knobs, like they are in channel mixer and Mic/Aux sections.
If you find any other issues please tell me in [Tango update PR](https://github.com/mixxxdj/mixxx/pull/1330)

ronso0 commented:
> I added an option to hide the EQ kill switches

Didn't test yet, but they are in a column with enable switch for QuickEffect. If you remove that as well it might lead to [Bug #1698408](https://bugs.launchpad.net/mixxx/+bug/1698408) if a users choses something else than filters.

Be-ing commented:
> > I added an option to hide the EQ kill switches

> Didn't test yet, but they are in a column with enable switch for QuickEffect. If you remove that as well it might lead to Bug #1698408 if a users choses something else than filters.

It's an option, so if that's an issue for a user they can leave the kill switches showing. They are showing by default.

ronso0 commented:
@daschuer 
> We can advertise Deere as a full feature skin. Tango and Schade can be focused on other use cases

I started Tango as a controller focused skin, but I think it contains all features now.
What exactly do you miss?

Be-ing commented:
I kinda feel like it's too perfectly symmetrical now...
![image](https://user-images.githubusercontent.com/9455094/30526577-39ece0dc-9be2-11e7-8e3a-2faa54cab1fa.png)
Thoughts?

ronso0 commented:
It looks well arranged and clear to me. Maybe a little more space in between the effect units
What's bothering you?

ronso0 commented:
> Find a better place for beatgrid editing buttons?

Maybe it helps if BPM label has some kind of hint that it's expandable, not only a hover effect?

Be-ing commented:
>  What's bothering you?

I'm not sure...

> Maybe a little more space in between the effect units

The effect units need that horizontal space when expanded.

> Maybe it helps if BPM label has some kind of hint that it's expandable, not only a hover effect?

Good idea.

ronso0 commented:
>  > Maybe a little more space in between the effect units

> The effect units need that horizontal space when expanded.

What I mean is a 4px gap like between mixer and decks, not a black hole like in Tango -_o

ronso0 commented:
With separate waveforms (not parallel) and without crossfader, mixer looks a bit tattered and non-existant alignment of level meters and volume faders becomes obvious.
Maybe expanding spacers above and/or below crossfader helps.

Be-ing commented:
Yeah, as mentioned in the TODO list at the top of the PR polishing the mixer alignment for all configurations has not been completed yet.

ronso0 commented:
> Find a better place for vinyl control buttons?

1. leave them next to waveforms, put them in a 2x2 grid
2. replace Eject button with deck-specific Vinyl Controls toggle, put them in 2x2 grid next to it

ronso0 commented:
Could you style cf assign buttons a way that they are eye-catching only when their position differs from default?
Can send you a small PR if you like

Be-ing commented:
@ronso0 a pull request would be appreciated.

Commit:
fix highlighted xfader orientation buttons

Commit:
push fx units 4px  apart, fix/improve spacer template

Be-ing committed:
Merge pull request #11 from ronso0/Deere-2.1-update2

fix highlighted xfader orientation buttons

Be-ing commented:
I will be really busy the next few weeks between school and moving to a new apartment. @ronso0 would you like to continue working on this?

ronso0 commented:
Sure, I could do this. We just need to agree how to proceed:
- Show recording & broadcasting icons when not active
What do you mean?

- Find a better place for vinyl control buttons?
**1** leave them next to waveforms, put them in a 2x2 grid, so that scrolling waveforms wouldn't be forced to such a great height anymore
**2** replace Eject button with deck-specific Vinyl Controls toggle, put them in 2x2 grid next to it
IMO **2** might shrink the waveforms too much if all mixer controls are enabled, so I'd implement **1**

- Find a better place for beatgrid editing buttons?
For now, they could stay next to pitch slider. BPM should be styled like any other button and stay rightmost when beatgrid controls are expanded

Commit:
always keep decks/mixer at minimum size

Commit:
re-order VinylControls, rename HOTCUE to HOT so it fits

Commit:
symmetric Mute buttons in 4-deck mixer

Be-ing commented:
> Show recording & broadcasting icons when not active
> What do you mean?

The icons should always be visible in an inactive state and lit up when active like any other button. This will make it harder to forget to press Record before starting a set.

> Find a better place for vinyl control buttons?

Option 1 seems like a good idea.

> BPM should be styled like any other button and stay rightmost when beatgrid controls are expanded

Sure, let's try it.

Commit:
fix position of mixer elements for all configurations

Commit:
left click focus button to unfocus

Commit:
always show Recording and Shoutcast button

Commit:
make BPM/Key controls more discoverable

Commit:
show/hide vinyl controls from within deck overview row

Commit:
resize Rec, Shoutcast & Vinyl icon to fit buttons

Commit:
stop Crossfader overlapping mixer container

Commit:
widen Crossfader background

Commit:
equalize padding of vinyl controls and FX assing buttons

Commit:
fill Spinny background in expanded stacked waveforms

naught101 commented:
Re: eq knob labels, what about a "Knob labels" option in the Mixer section in the skin settings?

Deere is a really nice skin, I've only just stated using it, but I really like it. I can see some other potential improvements for it though:

- The bottom section of each deck is rather empty, and could house way more stuff.
- The mixer section is really dense. Not great if you're actually using the UI to control things.
    - Moving the filter/super knob into right side of the the bottom section of the deck would make more space for labels, when they're on.
    - The PFL button could go below the meters, to give the volume fader more height.
    - This is way more work, but if you wanted to save horizontal space, the volume faders and meters could be merged into one control, like in Ableton (although probably with a more obvious fader position indicator). This would also make a tonne of sense visually, although it would map to physical controllers less well.
- The tempo fader is really short. The length of this is way more important than the length of the volume fader, IMO, because small changes are really noticable.
   - The Sync, Cue, and Play buttons could all be moved into the bottom section of the deck, to give the tempo fader more height. The pitch nudge should probably stay there.
- The parallel waveforms take up a lot of space. I would love it if there was a shorter version. Some options:
    - Just make them shorter. I have them as small as possible, and it's still really, really easy to read. It doesn't need to be that big.
    - exp(amplitude) scaled waveforms would make the beats stand out more, and make them easier to match.
    - Waveforms could probably happily overlap a little bit, without too much loss of usefulness. At the moment, I'm watching a track that's peaking at 80% on the meter, but it's only using 50% of the actual available space in the waveform.
   - It would be nice if the parallel waveforms didn't have to be the same format as the deck waveforms. The deck waveforms are for track overview, and getting an understanding of where you are in the track, especially in terms of energy. The parallel waveforms are primarity for beatmatching. VDJ handles this really well, using a super simplified exponential, colour-per-deck waveform for the parallel waveforms, making beatmatching simpler: 

![VDJ beatmatching](https://www.virtualdj.com/image/77814/107991/waveform.png)

- I find it visually weird that the loop controls just sit in the middle.. they feel like they're hanging in space. But I don't really have a good suggestion for how to avoid that. Perhaps just right-align them, with a visual bar between them and the beatjump controller?
- Why are there only 2 effects decks? There is plenty of space for 4. You could probably fit 4 in half the width..
- 8 sampler decks? That seems excessive. I can see how they've be useful for some people, but I guess most people would also be happy with 2-4. An option for fewer would be great. Then the Hotcues could also be moved up in to the main area of the sampler decks.

![20171019_134557](https://user-images.githubusercontent.com/167164/31758025-6f851876-b4f7-11e7-9c91-c5ad02c11b22.jpg)

- if the effects decks were half as wide (or a quarter as wide), then 4 sampler decks could all sit on their right, taking up less height in total.
- Mics and Aux also take up WAY too much space. They could almost be condensed up into the very top bar. 
- Personally I don't like the dark blue behind the crossfader. Can it be the same grey as the decks themselves?
- The skin as a whole could potentially benefit from some more contrast. In particular, using white-on-black sections could make important things stand out a bit more (this might be good for the top sections on each deck, with the title and artist). This would be a straight rip from Traktor though.

All that probably comes across as a bit harsh, but it's intended to be constructive. In general the
skin is fucking great, super professional, and easy to use, and I really appreciate the amount of effort that's been put in to it. If any of the above points need clarification, or if I can help with feedback on other design aspects, let me know. Rock on!

naught101 commented:
Whoops, corrected that screenshot.

naught101 commented:
Just noticed the "4 effects units" button in the settings. Ignore the comment about 2 effects units, but the comment about the width of the effects apply doubly, considering that the 4 effects units show up in 2 rows.

Be-ing commented:
> Re: eq knob labels, what about a "Knob labels" option in the Mixer section in the skin settings?

I think that would be overcomplicated and a waste of space. Tooltips are good enough for most of the rest of the skin so I don't know why they're not good enough for the mixer knobs.

> The bottom section of each deck is rather empty, and could house way more stuff.

I agree. Redesigning the decks to fix this and the short pitch faders would be nice, but a lot of work. If you or @ronso0 want to try it, please do.

> Moving the filter/super knob into right side of the the bottom section of the deck would make more space for labels, when they're on.

I don't like this idea. I've never seen any mixer or controller designed like this. The only DJ equipment I've seen do this is LateNight.

> The PFL button could go below the meters, to give the volume fader more height.

I like having the meters and volume fader the same height.

> This is way more work, but if you wanted to save horizontal space, the volume faders and meters could be merged into one control, like in Ableton (although probably with a more obvious fader position indicator). This would also make a tonne of sense visually, although it would map to physical controllers less well.

I have thought about this but I don't think it works out well. Traktor does it too and I find it difficult to see what is going on when the top of the meter and the fader are near the same height. The way Ableton does it would reduce the surface area available to grab the fader with the mouse or touch screen.

> Just make them shorter. I have them as small as possible, and it's still really, really easy to read. It doesn't need to be that big.

From your screenshot, it looks like you're using parallel waveforms with the spinny showing. Unfortunately, this requires setting an arbitrary minimum size to fit the spinny. If you disable the spinnies you can shrink it further. I attempted to fix this by making the spinnies resizable in #1079 but didn't figure it out. It would require some ugly complicated hacks on the C++ side that would not be easy to implement.

> Waveforms could probably happily overlap a little bit, without too much loss of usefulness. At the moment, I'm watching a track that's peaking at 80% on the meter, but it's only using 50% of the actual available space in the waveform.

Interesting idea, but I'm not sure I agree. I like seeing the RGB color coding of the frequency content for each track clearly separated. I think it would get difficult to tell which track is contributing which frequency content if they overlapped. That said, I'd be open to trying it if someone wanted to attempt it. This would require a lot of work to make a new combined waveform widget instead of two separate skin widgets.

> It would be nice if the parallel waveforms didn't have to be the same format as the deck waveforms.

They don't have to be. You can change this in the Preferences > Interface > Waveforms. Personally I find the RGB color coding of frequency content helpful for both. And it looks cool :)

> You could probably fit 4 [effects units] in half the width.. You could definitely fit 4 on a single row.

No, that's how they were in Deere 2.0 and the text and knobs were unreadably small.

> 8 sampler decks? That seems excessive.

Most controllers are designed for 4 or 8 samplers per side. The currently layout is designed to match that.

> Mics and Aux also take up WAY too much space. They could almost be condensed up into the very top bar.

Yeah, there is room for improvement there.

> Personally I don't like the dark blue behind the crossfader. Can it be the same grey as the decks themselves?

It is when AutoDJ isn't active. I'm not especially fond of the dark blue to indicate AutoDJ being active, so if you have a better suggestion, please share.

> The skin as a whole could potentially benefit from some more contrast. In particular, using white-on-black sections could make important things stand out a bit more (this might be good for the top sections on each deck, with the title and artist). This would be a straight rip from Traktor though.

Interesting idea.

Thank you for the detailed feedback!

ronso0 commented:
Thanks @naught101 for your feedback!

> This is way more work, but if you wanted to save horizontal space, the volume faders and meters could be merged into one control,

This could work if Volume handles had a transparent gap in the middle so that level is always visible.
Though would require to rework pretty much the whole mixer...

re: Waveforms
scrolling waveforms are not meant for beatmatching only, with RGB they're pretty useful to see what comes next (bass, hi-hat, vocals). In general, and if one has headphones plugged in, I recommend to learn beatmatching by ear step-by-step, waveforms might trick you.

naught101 commented:
@ronso0: Yeah, I know re: beatmatching. I do beatmatch by ear, but sometimes the waveforms help. Agree that the RGB display is super useful, too.

@Be-ing: Fair enough on most points.

- Auto DJ colour: the android icon in the top right is enough. The colour change is not an obvious indicator (I never would have guessed that, had you not told me). The icon could perhaps be placed somewhere in the mixer section though, to stand out more? Above or below the master meter, maybe?
- waveform height: The overview waveforms have a "normalise" option in the settings, that makes them take up all of the vertical space. Could that be applied to the summary waveforms as well? that would help a lot. (Also "summary" means the same thing as overview.. maybe they could be renamed to "proximal"? I can't find a better word for this..)
- re: EQ rack widths: What's the minimum screen size you're designing for? You can see in my screen shot that there is at least double the space required at 1080p. Agree that quartering it would be pushing things though.
- Re: super/filter knob. yeah, good point. I think the mixer is actually quite well laid out, although it is very tight in 2 deck mode.
    
Thanks for the detailed responses. If I do ever get time to start doing some work on Mixxx, I have a few other things I will attack before the Deere layout, but I am looking forward to seeing what you all come up with :)

Commit:
fix Master VU position for all configurations

Commit:
brighter border for BPM/Key edit triggers

Commit:
move Play/Cue to deck controls row, compress loop & beatjump

Commit:
fix play button: centered icon, 40-60px wide

Commit:
align loop_activate with reloop

Commit:
increase crossfade margin

Commit:
improve Rec & ShoutCast icons

Commit:
improve xFader highlight with AutoDJ enabled, sort toolbar indicators

ronso0 commented:
As I mentioned a while ago, and as @naught101 points out here, the knobs would appreciate a redesign, as well as the fader handles. For example we could change the white rings into a round, dark background to mark the knob range and make the knob indicator stand out more by increasing contrast locally.

Be-ing committed:
Merge pull request #12 from ronso0/Deere-2.1-update2

fix positions of mixer controls, rearrange vinyl controls

Commit:
compact Mic/Aux rack

Commit:
equalize major skin elements' spacing

Be-ing committed:
Merge pull request #13 from ronso0/deere_2.1_update_MicAux

improve Mic/Aux rack, equalize layout spacing

ronso0 commented:
@Be-ing I see there are some issues left and don't know if you want to solve them in this PR:
- Mic1 and Aux1 controls show up although no devices are configured
maybe show just Mic/Aux containers containing a text hint or a gear icon that would open device preferences, if that's possible?
- collapsed FX units take to much vertical space
a) implement a Tango-like toggle/EffectSelector combination (guess you don't like it..)
b) shrink EffectSelector and put Meta knob next to it, the row above contains enable button (and focus button)
- FX enable button is IMO too small considering its impact
- more elegant knobs and fader handles

Be-ing commented:
> Mic1 and Aux1 controls show up although no devices are configured
maybe show just Mic/Aux containers containing a text hint or a gear icon that would open device preferences, if that's possible?

That already happens when pressing the TALK button on the first mic. IIRC this was a conscious design decision in the past to show the user something so it didn't look like the mic/aux GUI just didn't work. But I like your idea of showing a more self-explanatory hint better. I think it would require creating a new skin widget to open the Sound Hardware Preferences.

> collapsed FX units take to much vertical space
a) implement a Tango-like toggle/EffectSelector combination (guess you don't like it..)
b) shrink EffectSelector and put Meta knob next to it, the row above contains enable button (and focus button)
FX enable button is IMO too small considering its impact

I think the effects units are good enough as they are, but if you want to try improving them further, go for it. I don't really like the combination of the effect enable switch and selector widget. Shrinking the effect selector a bit might work with a bit smaller font size.

> more elegant knobs and fader handles

I don't see this as a pressing need, but if you want to propose some better graphics I'm open to ideas.

> more distinct knob labels in top row: Head Gain, Master Gain, ...
to keep them narrow, each could be split into two lines

Remember they have to fit in the minimum width. It would be nice if the word "Gain" could fit up there where appropriate, but I don't think it's a big deal with tooltips. I'm not in favor of using any more vertical space for those.

ronso0 commented:
> That already happens when pressing the TALK button on the first mic. IIRC this was a conscious design decision in the past to show the user something so it didn't look like the mic/aux GUI just didn't work

Ah okay, didn't try that, makes perfect sense.

> Shrinking the effect selector a bit might work with a bit smaller font size.

For me the issue is the two-line layout that eats to much space, especially on small screens. Decreasing font size alone won't help much, and the padding incl. preserved space for the focus border makes it coherent with the rest of the skin.

Just feel that Deere as a whole could be more harmonious with less toy-like knobs and faders, but if it's working for the majority as it is I don't feel the urge to spend time on it.

nopeppermint commented:
I like Deere as It is at the moment, and not smaller.

sblaisot commented:
I suppose left char on skin menu should be tick marks or something to distinguish when activated. This is not working on windows:
![image](https://user-images.githubusercontent.com/2768193/32199530-3262f560-bdcd-11e7-8d1a-7444538e131d.png)

ronso0 commented:
> I like Deere as It is at the moment, and not smaller.

That really depends on the screen size and the configuration you use. Deere --like Shade and Tango-- claims to fit netbook screens (1024x600) and then it gets quite challenging..

ronso0 commented:
> I suppose left char on skin menu should be tick marks or something to distinguish when activated.

Those characters (empty and filled circles) are actually in the Open Sans char table. The font looks a bit different in your screen shot, i.e. the capital C and the height of all small letters.
Was it always like that? Could you check if `/res/fonts/` contains OpenSans-Light.ttf and OpenSans-Regular.ttf

sblaisot commented:
I remember in the past the empty/filled circle was working right.

there is no res/fonts (that's a source directory) but I have a C:\programs\Mixxx\fonts directory containing OpenSans-Light.ttf and OpenSans-Regular.ttf

BUT they are not installed in the system, so Mixxx has to explicitely load them. is this working ? should they be installed system-wide ? (in which case I need to update the installer)

sblaisot commented:
Deere before this PR show the same behaviour (as found in master right now).

sblaisot commented:
installing these fonts system-wide donesn't change anything. still no circle.

Be-ing commented:
>  I will try to dissect tomorrow and see which commit added this regression.

I changed the characters used in the skin menu back in #940. The question now is how to get the intended font to load on Windows.

sblaisot commented:
should be already loaded:
https://github.com/mixxxdj/mixxx/blob/master/src/util/font.h#L13
https://github.com/mixxxdj/mixxx/blob/master/src/mixxx.cpp#L175

sblaisot commented:
starting in developer mode show the below log so the font seems to be properly loaded:
```
Debug [Main]: Loading resources from  "C:/Program Files/Mixxx/" 
Debug [Main]: FONT LOADED family: "Open Sans Light" style: "Maigre" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72" 
Debug [Main]: FONT LOADED family: "Open Sans Light" style: "Gras" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72" 
Debug [Main]: FONT LOADED family: "Open Sans Light" style: "Maigre Italique" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72" 
Debug [Main]: FONT LOADED family: "Open Sans Light" style: "Gras Italique" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72" 
Debug [Main]: FONT LOADED family: "Open Sans" style: "Normal" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72" 
Debug [Main]: FONT LOADED family: "Open Sans" style: "Gras" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72" 
Debug [Main]: FONT LOADED family: "Open Sans" style: "Italique" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72" 
Debug [Main]: FONT LOADED family: "Open Sans" style: "Gras Italique" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72" 
Debug [Main]: FONT LOADED family: "Ubuntu" style: "Gras" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72" 
Debug [Main]: FONT LOADED family: "Ubuntu" style: "Gras Italique" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72" 
Debug [Main]: FONT LOADED family: "Ubuntu" style: "Normal" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72" 
Debug [Main]: FONT LOADED family: "Ubuntu" style: "Gras" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72" 
Debug [Main]: FONT LOADED family: "Ubuntu" style: "Italique" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72" 
Debug [Main]: FONT LOADED family: "Ubuntu" style: "Gras Italique" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72" 
```

sblaisot commented:
ok, you used `&#9898` and `&#9899` which is respectively in hex `0x26AA`and `0x26AB`.
Unfortunately, `Open Sans` doesn't have any character between `0x25CA` and `0xFB00` so probably your system fallback to another font.

see http://www.fontspace.com/steve-matteson/open-sans/38455/charmap

so either you have to use a character that is included in this font or use another font.

Be-ing commented:
Thanks for looking into it. Do you have any hints for figuring out what font my system is falling back to? Or any alternative ideas for a character to use?

sblaisot commented:
I tried replacing "off" by a space and "on" by "&#215;" (multiplication sign, a little cross), it works but it's not as sexy as the filled circle.

![image](https://user-images.githubusercontent.com/2768193/32231392-baeb33f6-be55-11e7-9d69-3c3a1dd6dca9.png)

Unfortunately "Open Sans" doesn't have any tick mark or something that can reflect "selected".

I have no idea what font your system uses as a fallback.

sblaisot commented:
Some questions/comments about Deere 2.1: (no real blocker for merging, just a wishlist)
* skin text translations: how do you plan to translate skin text like the skin settings menu?
* Parallel waveforms spinnies have a fixed minimum size preventing to narrow parallel waveforms. You can narrow them when you disable spinnies. Is it possible to remove the fixed minimum size of spinnies or at least have it smaller ?
* I miss sampler gain
* I miss sampler "headphones" too (to prelisten samples), although this one can be achieved with preview deck (yes, I also have a samplers volume on my controller)
* I find mic on left and aux on right a bit disturbing. couldn't we group them in the middle ? or at least center the mic section in the left half of the screen and the aux section in the right half ?
* when the screen is large enough, there is a blank gap between the first 4 samplers on the left and the remaining 4 on the right. Couldn't this blank be minimal (say a couple of pixels) to allow some more chars in the sample title of each sampler ? I'm aware that it is really hard to be perfect at the same time for small and large screens...

Be-ing commented:
> skin text translations: how do you plan to translate skin text like the skin settings menu?

We currently do not have a mechanism for this, but we should. I don't think it would be too difficult to implement.

> Parallel waveforms spinnies have a fixed minimum size preventing to narrow parallel waveforms. You can narrow them when you disable spinnies. Is it possible to remove the fixed minimum size of spinnies or at least have it smaller ?

Unfortunately some arbitrary size must be imposed because the spinnies cannot be resized. I attempted to get them to resize with #1079 without success.

> I find mic on left and aux on right a bit disturbing. couldn't we group them in the middle ? or at least center the mic section in the left half of the screen and the aux section in the right half ?

Sure, if you want to implement that. I think it would be fine either way.

> I miss sampler gain
> I miss sampler "headphones" too (to prelisten samples), although this one can be achieved with preview deck (yes, I also have a samplers volume on my controller)

Every widget added increases startup and shutdown time because it is instantiated 64 times. We need to be judicious in what is shown in samplers. In the long term, I think we should replace the current implementation of samplers as simply renamed decks with a sampler/remix decks like Traktor that mix a group of samples together into a group that can be manipulated with normal deck controls (volume fader, PFL switch, EQs, effects).

> when the screen is large enough, there is a blank gap between the first 4 samplers on the left and the remaining 4 on the right. Couldn't this blank be minimal (say a couple of pixels) to allow some more chars in the sample title of each sampler ? I'm aware that it is really hard to be perfect at the same time for small and large screens...

Yeah, making it work on low resolution and high resolution screens is difficult. Pull requests welcome.

ronso0 commented:
> Unfortunately, Open Sans doesn't have any character between 0x25CA and 0xFB00 so probably your system fallback to another font.

ups, I was pretty sure, remebering the discussion in #940 or earlier

ronso0 commented:
> Yeah, making it work on low resolution and high resolution screens is difficult. Pull requests welcome.

Just replace `<SizePolicy>me,min</SizePolicy>` with `<Size>10f,0me</Size>` (or whatever size you prefer) [sampler_row.xml:line42](https://github.com/mixxxdj/mixxx/pull/1343/files#diff-7cf1df29f73ac1e5f5dc587e0912bed8R42)

Commit:
fix button grid in deck overview row

ronso0 commented:
- highlighted rows in Control Panel or Tracks Table aren't readable if unfocused
(Ubuntu 14.04, GTK theme 'Ambiance' with custom icons)
![screenshot - 01 11 2017](https://user-images.githubusercontent.com/5934199/32276028-cbce4506-bf0d-11e7-9071-8a41df7231a6.png)

- minimum size should be 1008,xx to fit 1024px screens

uklotzde commented:
Very clear, functional, and compact skin. I would like to have this merged ASAP ;) Unfortunately I cannot provide any new insights as a Fedora user with the standard GNOME theme.

Commit:
fix some layout issues in deck controls & deck overview

Commit:
fix minimum skin size

Commit:
simplify container bg colors

Commit:
fix colors for selected library row

nopeppermint commented:
![linux_mint_18 2](https://user-images.githubusercontent.com/3403218/32702033-d5806cd8-c7e0-11e7-8e9f-bc618e2767c9.png)
Linux Mint 18.2 with Cinnamon 3.4.6 , 1920x1080 works without any error

Commit:
define colors for library Preview button

ronso0 commented:
I tried some alternate designs for the knobs, please check the 100% views to judge.

segmented ring
![deere_knobs_1](https://user-images.githubusercontent.com/5934199/32703534-c56edcfa-c7f7-11e7-9125-431b06d9dfcf.png)

dots
![deere_knobs_3](https://user-images.githubusercontent.com/5934199/32703535-c80580b8-c7f7-11e7-868b-5200950e8174.png)

dark ring
![deere_knobs_7](https://user-images.githubusercontent.com/5934199/32703536-c9b8ef30-c7f7-11e7-90ca-deb23be6f329.png)

My favorite is definitely the dark ring, it matches the current slider notch design the most.

Be-ing commented:
Hmm, aesthetically I'm not too keen on any of those knob proposals, but there are some good ideas there. I like how the dotted ring highlights the center point. I like how the notch stands out more from the white in the center on all of them, but I think it stands out a little too much.

poelzi commented:
I go with @Be-Ing that an indicator for center is very good.
Personally I like the second the most, even better then the current design.

Be-ing commented:
How about putting a small center notch of the outer ring of the current design and making the position marker a little more prominent?

nopeppermint commented:
I prefer the segmented ring, the dark ring is hard to see, the white one is better

ronso0 commented:
Yes, the indicator could be a bit shorter.
IMO the outer white ring makes no sense, even with a notch. I just let my flatmates test it: with a white ring and yellow indicator (even if it stands out) it's all just a whitish stain from a meter away.
So either we find a better color than yellow, or we should use a darker scale. Though, with a dark scale all indicator colors work better.

ronso0 commented:
Yet another idea ;)

![deere_knobs_9_colorrings](https://user-images.githubusercontent.com/5934199/32705985-f8ff1cca-c819-11e7-9452-5a7edd5ae258.png)

naught101 commented:
I like that last design - it looks much nicer with the smaller FX parameter knobs. Also the orange is better than the yellow.

Would it be possible to do a variant of that with the marker and marker outline a little thicker (so more similar to the small knobs, proportionally, and also include the marker outline over the top of the coloured outline ring (so there is more separation between the marker and the ring)?

Be-ing commented:
Interesting. I don't think there's enough contrast between the inside of the knobs and the notch to tell them apart very easily.

ronso0 commented:
Couldn't sleep and gave it another try. Funny that it matches what you commented in the meantime.
I also tried to change colors so `red` always refers to some kind of volume, and `blue` is for effects.
Check it out. The center notch can be improved.

![deere_knobs_9_colorrings fader2](https://user-images.githubusercontent.com/5934199/32708567-f1fb6a8a-c82a-11e7-8774-dfd3f4ca96cb.png)

naught101 commented:
Looking good. Maybe just make the outline two lines broken at the centre, instead of with a grey marker so the ends are curved? And maybe a tiny bit further apart, so the gap is bigger.

Also, maybe make the indicator lines slightly shorter on the big knobs, and slightly longer on the small knobs?

ronso0 commented:
Sure! These are quick shots to test your opinions. If we have a consensus I'll go ahead and waste time on the details

Be-ing commented:
How about making the inner area of the knobs a bit darker? Also, any ideas for the small effect parameter knobs?

ronso0 commented:
Even darker? I'll try.
I did't plan changing the shape of the small knobs, I think they're okay. They use every pixel available and color doesn't add any value. Contrast could be higher though.
Hmmaybe that circle could be a ring and inner area is filled 280/360° to mark the limits and center point, but a clear design is tricky since they're really tiny.
How do you like the fader handles?

poelzi commented:
@ronso0  I really like the last version. could You make the center dot more visible ? On high res monitors it's very faint.

ronso0 commented:
I'm glad you like it! To test up-to-date version just
`git merge https://github.com/ronso0/mixxx/tree/deere_2.1_knobs_redesign` 
and `git merge https://github.com/ronso0/mixxx/deere_2.1_layout-fixes` to test layout fixes as well.
PR is here [knobs redesign #16](https://github.com/Be-ing/mixxx/pull/16)

![deere_knobs_9_colorrings fader3](https://user-images.githubusercontent.com/5934199/32730414-4f34a78c-c887-11e7-8814-338b6d505acc.png)

Commit:
simplify skin.xml to minimize flickering when togling stacked waveforms

ronso0 commented:
I experience that the spacing via `qproperty-layoutSpacing` is not working reliably in Deere, instead several button grids look like this again, although we already fixed it:
![grid_1](https://user-images.githubusercontent.com/5934199/32752293-ee3f050c-c8c8-11e7-8ce3-4933c3ebeb10.png)
When I restart mixxx multiple times while testing the padding in between the buttons sometimes even vanishes completely. I had headaches suspecting I messed up my git branches. Git is fine but the problem persist.
Can anyone confirm this?

ronso0 commented:
Like @sblaisot I miss sampler gain. A knob in the top bar might be sufficicent, visible only when samplers are enabled.

Be-ing commented:
I'm not pushing any more commits to this PR until I get write access. This is ridiculous that this branch has been sitting ready to merge for 1.5 months and pull requests are being targeted at my fork instead of master. The discussion has strayed far off topic from the original purpose of this PR.

daschuer commented:
Sorry for not following all details of this branch an merge it in a time. @ronso0: Is it merge-able in the current state? Do we have a finished review with a formal LGTM yet? 

@Be-ing: There is still an open TODO in the description, do you plan to fix it?  I do not understand how master write access will change the situation? We have discussed that no one should merge his own PRs without a formal LGTM. Thats why #894 is for example also not merged. Accepting PRs to private branches is IMHO a good technique to push a branch forward, that is not in a mergable state.

Be-ing commented:
> Sorry for not following all details of this branch an merge it in a time.

It's not your fault. It should not be any one person's responsibility to follow every pull request in detail.

> There is still an open TODO in the description, do you plan to fix it?

Yes, I do. The master level meter with 4 decks parallel waveforms is not the same height as the deck meters, but I don't think that's necessary to fix before merging. We could go on forever improving this but it has been in a good mergeable state that is better than master for weeks. For the future let's separate TODO lists into tasks that must be fixed before merge and those that don't need to be fixed before merge so this is clearer.

> We have discussed that no one should merge his own PRs without a formal LGTM. Thats why #894 is for example also not merged. Accepting PRs to private branches is IMHO a good technique to push a branch forward, that is not in a mergable state.

I agree that we generally shouldn't merge our own PRs. @ronso0 and @uklotzde have tested this and should have been able to merge it too. PRs for private branches are good for branches that aren't in condition to merge to master, but this has been for a while.

ronso0 commented:
@Be-ing A while ago you asked me to help you fix Deere because you expexted to have little time for it.
I helped out and opened PRs against this (your) branch since you somehow took over design and UX leadership for Deere with the previous Deere update PR. That -and only that- is the reason why I opened the PRs against your fork rather than master so you can test them and merge with this PR if you find the changes worth. There's still one ToDo open in the OP of this PR and you didn't ask for final approval of the changes done so far. Furthermore you joined the discussion about redesigning the knobs I started here. If you don't like that procedure, please let me know earlier.

ronso0 commented:
I think for skin PRs like this we should generally make sure that they were tested on all platforms before merge, maybe by checkboxes in the OP. See previous glitches in EffectSelector and beat spinboxes.

Be-ing commented:
@ronso0, you've done nothing wrong by sending me PRs and I apologize if it came off that way. My point is this branch should've been merged to master weeks ago and your recent PRs should have been able to be targeted at master.

Be-ing commented:
> I think for skin PRs like this we should generally make sure that they were tested on all platforms before merge, maybe by checkboxes in the OP. See previous glitches in EffectSelector and beat spinboxes.

That would be ideal, but I don't think we have enough people testing on every platform (especially macOS) to make that a requirement for merging a PR. It should certainly be a requirement for a release though.

ronso0 commented:
IMO this is ready to merge. On Linux it's fine, @sblaisot tested it for Windows I suppose. Did anyone do so for OSx?
@Be-ing  If we want to fix the remaining issues I'd like [my layout fixes PR](https://github.com/Be-ing/mixxx/pull/15) to be tested and merged, instead of patchworking with PRs for every little skin issue.

I'll re-open the knob redesign PR against master. Right now it seems to be a consensus to drop LateNight and Shade as long as they offer neither the effects nor the looping UI. So it seems very likely Deere will be the default skin and I think we can do better than what we have in the current design.

Be-ing commented:
I did not explicitly ask for merging this before because I am exhausted from having to repeatedly ask one person for every little change. This cannot be one person's responsibility if we are going to have releases more often than every two years. We have "continuous integration" services, but what we practice is the opposite of continuous integration.

ronso0 commented:
Okay, I somehow understand your frustration and it's great you started that email discussion about it.
For now we have to live with it, I guess. No matter how many members are able to merge, skin PRs like this are practically impossible to safely review code-wise, members have to rely on trusted reviewers to test as many skin configurations as possible on various OS. No big deal to request LGTMs from users and then ping @daschuer to merge.

Pegasus-RPG commented:
@esbrandt Are you able to take a look at this?

Be-ing committed:
Merge pull request #15 from ronso0/deere_2.1_layout-fixes

Deere 2.1 layout & color fixes

uklotzde commented:
@Be-ing There is still one open task on your check list. Do you plan to finish it here? Otherwise I would like to merge this PR and the polishing can be done afterwards. We already have #1399 as a follow up.

Great job 👍 The UI of Mixxx is evolving at an impressive speed. I hope we can pick up this momentum for other parts of the application, too! We definitely need more domain experts in all areas.

ronso0 commented:
@uklotzde I fixed the mixer layout for all configurations.
Quoting @Be-ing here the only tiny 'issue' left is
> The master level meter with 4 decks parallel waveforms is not the same height as the deck meters, but I don't think that's necessary to fix before merging.

I think this is ready to merge, too.

uklotzde commented:
The channel meters are tiny in the full featured 4-deck parallel layout, anyway. Having a bigger master level meter instead of wasting space might be desirable, even if it doesn't match the scaling of the channel meters. We might find a compromise or even new ideas for different options afterwards.

LGTM.

uklotzde commented:
Merging this now since no one voted against it. The beta release date is near and we should get feedback from some alpha testers about our main skin.

naught101 commented:
New interface is 👌 

Thanks so much for your work @Be, @ronso0, and others!

sblaisot commented:
Deere 2.1 menu has not been corrected. Still squares.
![image](https://user-images.githubusercontent.com/2768193/34232813-6deb548c-e5e2-11e7-8a34-1c07c62b5919.png)

Be-ing commented:
Pull requests welcome.

JosepMaJAZ commented:
I didn't make the new pull request to fix the square boxes because of the
other problem with the samplers that I don't know how to fix. Anyway, I'll
repeat here what's the problem: the unicode characters used in the skin for
these graphics does not exist in the Windows fonts. There are other codes
with similar characters that are present on both, Linux and Windows, but I
don't have it here right now.

sblaisot commented:
Well. the problem is bigger than that. We explicitely load a font. so we shouldn't rely on fallback font using characters that are not embedded in these fonts.

for me, we have 2 solutions:
1. Use characters that are included in the font. I tried but didn't find suitable characters
2. change font

sblaisot commented:
> Pull requests welcome.

I know that :)
that was more an informational message

JosepMaJAZ commented:
I found the email I sent to be. As for loading a font, is Mixxx embedding
fonts?  Afaik, it only uses installed fonts.


I finally found out what happens, which is that the characters that you
use, are not recognized on Windows.
I found replacements that work on both, Windows and Linux (I even tried
under Windows 7):
Here, the ones you use (in HEX)
https://www.fileformat.info/info/unicode/char/26AA/index.htm
https://www.fileformat.info/info/unicode/char/26AB/index.htm

Here, the compatible ones (in HEX)
https://www.fileformat.info/info/unicode/char/25CB/index.htm
https://www.fileformat.info/info/unicode/char/25CF/index.htm

Decimal values correspond to 9675 and 9679

sblaisot commented:
Yeah, that works ! Thanks. PR #1430
